design-notes/API.md

```diff
@@ -0,0 +1,135 @@
+
+
+## Gaussian approximated methods
+
+### Model definition
+
+The main user defined function will be the way to compute (or approximate) conditional mean and covariances
+
+```python
+@cholesky_model
+def function(params_at_time_t):
+    # Do stuff with the params
+    
+    def m(x_t):
+        """ Computes :math:`\mathbb{E}[x_{t+1} \mid x_t]`
+        """
+        return ...
+
+    def chol(x_t):
+        """ Computes :math:`\mathbb{C}[x_{t+1} \mid x_t]^{1/2}`
+        """
+        return ...
+    return m, chol
+
+# Otherwise, to allow for logic to be shared, we can also do 
+
+@cholesky_model
+def function(params_at_time_t):
+    # Do stuff with the params
+    
+    def m_chol(x_t):
+        """ Computes :math:`\mathbb{E}[x_{t+1} \mid x_t]` and :math:`\mathbb{C}[x_{t+1} \mid x_t]^{1/2}`
+        """
+        return MVNChol(...)
+
+    return m_chol
+
+```
+
+where for instance the `cholesky_model` decorator can be defined as 
+
+```python
+class CholeskyModel(NamedTuple):
+    mean: Callable
+    chol: Union[Callable, np.ndarray]
+    is_chol_constant: bool
+    
+def cholesky_model(function):
+    @functool.wraps(function)
+    def get_model(*args, **kwargs):
+        m, chol = function(*args, **kwargs)
+        return CholeskyModel(m, chol, not iscallable(chol))
+    return get_model
+```
+
+similarly for covariance models
+
+```python
+class CovarianceModel(NamedTuple):
+    mean: Callable
+    cov: Union[Callable, np.ndarray]
+    is_cov_constant: bool
+    
+def covariance_model(function):
+    @functool.wraps(function)
+    def get_model(*args, **kwargs):
+        m, cov = function(*args, **kwargs)
+        return CovarianceModel(m, cov, not iscallable(cov))
+    return get_model
+```
+
+Of course, in some cases, the conditional mean and covariance is not tractable in closed-form so that we need to provide the user with a utility function that takes a complicated (functional at least) model, and returns the (appoximated) conditonal one.
+
+```python
+def make_conditional_model(f: Callable[[np.ndarray, np.ndarray, ...], np.ndarray], integration_method, sqrt:bool) -> Union[CovarianceModel, CholeskyModel]:
+    # For example here we suppose that `q` is a MVNCov
+    if sqrt:
+        @covariance_model
+        def mean_cov(m_t, cov_t, *args, **kwargs):
+            def mean(x_t):
+                # Something like this I guess, but we can iterate.
+                F, bias, _ = linearize(lambda q_: f(x_t, q_, *args, **kwargs), integration_method, MVNCov(m_t, cov_t))
+                return F @ m_t + bias
+```
+
+Finally, the user can also pass in LGSSMs directly to a public Kalman API beside the non-linear extensions. This will be done in a completely explicit way, where the user will feed in all matrices and say if they want the sqrt or standard form. That is,
+
+```python
+def kalman_filter(ys: np.ndarray, F: np.ndarray, Q: Union[MVNCov, MVNChol], H, R:Union[MVNCov, MVNChol], x0: Union[MVNCov, MVNChol], propagate_first:bool = False):
```

Comment:
We should allow all the parameters to be optionally parameterized by time, to handle non-stationary models.
This is what we did in https://github.com/probml/JSL/tree/main/jsl/lds:

```
@dataclass
class LDS:
    A: chex.Array
    C: Union[chex.Array, Callable]
    Q: chex.Array
    R: chex.Array
    mu0: chex.Array
    Sigma0: chex.Array

    def observation_matrix(self, t: int):
        if callable(self.C):
            return self.C(t)
        else:
            return self.C


def kf(params, obs, ...):
  for t in range(T):
       Ct = params.observations(t)


```
I know you don't like OOP, but the above is still functionally pure, since dataclass is immutable.